app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/LoginViewModel.kt

```diff
@@ -0,0 +1,92 @@
+package com.sci.recipeandroid.feature.auth.ui.viewmodel
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.sci.recipeandroid.feature.auth.domain.repository.AuthRepository
+import com.sci.recipeandroid.util.SingleLiveEvent
+import com.sci.recipeandroid.util.Validator
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+class LoginViewModel(
+    private val authRepository: AuthRepository,
+    private val validator: Validator
+) : ViewModel() {
+
+    private val _loginFormState = MutableLiveData(LoginFormState())
+    val loginFormState: LiveData<LoginFormState> = _loginFormState
+
+    private val _loginScreenEvent = SingleLiveEvent<LoginScreenEvent>()
+    val loginScreenEvent: LiveData<LoginScreenEvent> = _loginScreenEvent
+
+
+    fun onEvent(loginFormEvent: LoginFormEvent) {
+        when (loginFormEvent) {
+            is LoginFormEvent.EmailChanged -> {
+                _loginFormState.value = _loginFormState.value?.copy(email = loginFormEvent.email)
+            }
+
+            is LoginFormEvent.PasswordChange -> {
+                _loginFormState.value =
+                    _loginFormState.value?.copy(password = loginFormEvent.password)
+            }
+            is LoginFormEvent.Submit -> {
+                _loginFormState.value?.let {
+                    val emailResult = validator.emailValidator(it.email)
+                    val passwordResult = it.password.isNotBlank()
+
+                    if (emailResult.successful && passwordResult) {
+                        logIn(it.email, it.password)
+                    }else{
+                        _loginFormState.value = _loginFormState.value?.copy(
+                            emailError = emailResult.errorMessage,
+                        )
+                    }
+                }
+            }
+        }
+    }
+
+    private fun logIn(email: String, password: String){
+        viewModelScope.launch(Dispatchers.IO) {
```

Comment:
IO handling

Reply:
moved this io handling at the datasource 9baf80468f57d78d6c4b13f545e1ebaa10f9d59c